In [2]:
import requests
import pandas as pd
import numpy as np
from pandas import json_normalize
from collections import ChainMap

In [4]:
def obtenerUrl(x):
    url = "https://opendata.aemet.es/opendata/api/prediccion/especifica/municipio/diaria/{}".format(x)
    querystring = {"api_key":"eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJyYWZvbnNvc3Bpbm9sYUBnbWFpbC5jb20iLCJqdGkiOiJmNDhmMmIyMS00ZTU1LTQ2NjgtYTdjZS1lNmY5M2RkYmU3NDciLCJpc3MiOiJBRU1FVCIsImlhdCI6MTYyNDAzODE5NSwidXNlcklkIjoiZjQ4ZjJiMjEtNGU1NS00NjY4LWE3Y2UtZTZmOTNkZGJlNzQ3Iiwicm9sZSI6IiJ9.RNfRWDVTw488ARtSjfihHdulZCo46n1IFnGmWNCsWHU"}
    headers = {
    'cache-control': "no-cache"
        }

    response = requests.request("GET", url, headers=headers, params=querystring)

    import json
    p = response.text
    d = json.loads(p)
    t = d.get('datos')
    return t

In [5]:
def obtenerDatos(x):
    r = requests.get(x)
    j = r.json()
    df = pd.DataFrame(j)
    return df

## Crear dataset de municipios con el código de AEMET  


In [37]:
df_munic_tot = pd.read_csv("C:/Users/saulo/Documents/R/data/MUNICIPIOS.csv", sep=";", engine='python')

In [38]:
df_munic_tot[df_munic_tot['COD_INE'] == 35016000000]

COD_INE   ID_REL  COD_GEO  COD_PROV   PROVINCIA  \
5265  35016000000  1350167    35001        35  Las Palmas   

                   NOMBRE_ACTUAL  POBLACION_MUNI SUPERFICIE  PERIMETRO  \
5265  Las Palmas de Gran Canaria          381223   10331,59     108650   

      COD_INE_CAPITAL                     CAPITAL  POBLACION_CAPITAL  \
5265      35016002401  Las Palmas de Gran Canaria             301685   

     HOJA_MTN25_ETRS89 LONGITUD_ETRS89 LATITUD_ETRS89 ORIGENCOOR ALTITUD  \
5265            1098-4    -15,41336841    28,09937855       Mapa       6   

     ORIGENALTITUD  
5265           MDT

In [39]:
df_munic_ine = pd.read_excel("C:/Users/saulo/Documents/R/data/20codmun.xlsx", skiprows=[0], dtype=str)

In [40]:
df_munic_ine[df_munic_ine['NOMBRE '] == 'Palmas de Gran Canaria, Las']

CODAUTO CPRO CMUN DC                      NOMBRE 
5266      05   35  016  7  Palmas de Gran Canaria, Las

In [41]:
df_munic_ine = df_munic_ine.rename(columns = {'NOMBRE ': 'NOMBRE'})

In [42]:
df_munic_tot["COD_INE"].dtype

dtype('int64')

In [43]:
df_munic_tot["COD_INE"] = df_munic_tot["COD_INE"].astype(str)

In [44]:
# Quito los 0 del final de la columna COD_INE, el problema es que me quita todos los 0 del final, no sólo 6
df_munic_tot["COD_AEMET"] = df_munic_tot["COD_INE"].str.strip("000000")

In [45]:
df_munic_tot.columns.values

array(['COD_INE', 'ID_REL', 'COD_GEO', 'COD_PROV', 'PROVINCIA',
       'NOMBRE_ACTUAL', 'POBLACION_MUNI', 'SUPERFICIE', 'PERIMETRO',
       'COD_INE_CAPITAL', 'CAPITAL', 'POBLACION_CAPITAL',
       'HOJA_MTN25_ETRS89', 'LONGITUD_ETRS89', 'LATITUD_ETRS89',
       'ORIGENCOOR', 'ALTITUD', 'ORIGENALTITUD', 'COD_AEMET'],
      dtype=object)

In [46]:
df_munic_tot = df_munic_tot[['COD_INE', 'ID_REL', 'COD_GEO', 'COD_AEMET', 'COD_PROV', 'PROVINCIA', 'NOMBRE_ACTUAL',
                             'POBLACION_MUNI', 'SUPERFICIE', 'PERIMETRO', 'COD_INE_CAPITAL', 'CAPITAL',
                             'POBLACION_CAPITAL', 'HOJA_MTN25_ETRS89', 'LONGITUD_ETRS89', 'LATITUD_ETRS89',
                             'ORIGENCOOR', 'ALTITUD', 'ORIGENALTITUD']]

In [47]:
df_munic_tot.head()

COD_INE   ID_REL  COD_GEO COD_AEMET  COD_PROV    PROVINCIA  \
0  1001000000  1010014     1010      1001         1  Araba/Álava   
1  1002000000  1010029     1020      1002         1  Araba/Álava   
2  1003000000  1010035     1030      1003         1  Araba/Álava   
3  1004000000  1010040     1040      1004         1  Araba/Álava   
4  1006000000  1010066     1060      1006         1  Araba/Álava   

      NOMBRE_ACTUAL  POBLACION_MUNI SUPERFICIE  PERIMETRO  COD_INE_CAPITAL  \
0  Alegría-Dulantzi            2935  1994,5872      35069       1001000101   
1           Amurrio           10264    9649,53      65652       1002000201   
2           Aramaio            1442    7325,84      41775       1003000601   
3        Artziniega            1800    2728,73      22886       1004000101   
4           Armiñón             223    1297,27      24707       1006000101   

            CAPITAL  POBLACION_CAPITAL HOJA_MTN25_ETRS89 LONGITUD_ETRS89  \
0  Alegría-Dulantzi               2823            0113-3     -2,51243731   
1           Amurrio               9188            0086-4     -3,00007326   
2            Ibarra                787            0087-4     -2,56540037   
3        Artziniega               1675            0086-1     -3,12791718   
4           Armiñón                100            0137-4     -2,87183475   

  LATITUD_ETRS89 ORIGENCOOR ALTITUD ORIGENALTITUD  
0    42,83981158       Mapa     568           MDT  
1    43,05427776       Mapa     219           MDT  
2    43,05119653       Mapa     333           MDT  
3    43,12084358       Mapa     210           MDT  
4    42,72326199       Mapa     467           MDT

In [48]:
len(df_munic_tot["COD_INE"])

8131

In [49]:
# Bucle para rellenar los 0 de la columna COD_AEMET y así obtener el código del INE que usa AEMET
for i in range(len(df_munic_tot["COD_AEMET"])):
    if ((len(df_munic_tot.iloc[i, 3]) == 3) & (len(df_munic_tot.iloc[i, 0]) == 10)):
        df_munic_tot.iloc[i, 3] = "0" + df_munic_tot.iloc[i, 3] + "0"        
    
    elif ((len(df_munic_tot.iloc[i, 3]) == 4) & (len(df_munic_tot.iloc[i, 0]) == 10)):
        df_munic_tot.iloc[i, 3] = "0" + df_munic_tot.iloc[i, 3]

    elif ((len(df_munic_tot.iloc[i, 3]) == 2) & (len(df_munic_tot.iloc[i, 0]) == 10)):
        df_munic_tot.iloc[i, 3] = "0" + df_munic_tot.iloc[i, 3] + "00"

    elif ((len(df_munic_tot.iloc[i, 3]) == 3) & (len(df_munic_tot.iloc[i, 0]) == 11)):
        df_munic_tot.iloc[i, 3] = df_munic_tot.iloc[i, 3] + "00" 
        
    elif ((len(df_munic_tot.iloc[i, 3]) == 4) & (len(df_munic_tot.iloc[i, 0]) == 11)):
        df_munic_tot.iloc[i, 3] = df_munic_tot.iloc[i, 3] + "0"          

In [50]:
df_munic_tot.shape

(8131, 19)

In [51]:
df_munic_tot.head()

COD_INE   ID_REL  COD_GEO COD_AEMET  COD_PROV    PROVINCIA  \
0  1001000000  1010014     1010     01001         1  Araba/Álava   
1  1002000000  1010029     1020     01002         1  Araba/Álava   
2  1003000000  1010035     1030     01003         1  Araba/Álava   
3  1004000000  1010040     1040     01004         1  Araba/Álava   
4  1006000000  1010066     1060     01006         1  Araba/Álava   

      NOMBRE_ACTUAL  POBLACION_MUNI SUPERFICIE  PERIMETRO  COD_INE_CAPITAL  \
0  Alegría-Dulantzi            2935  1994,5872      35069       1001000101   
1           Amurrio           10264    9649,53      65652       1002000201   
2           Aramaio            1442    7325,84      41775       1003000601   
3        Artziniega            1800    2728,73      22886       1004000101   
4           Armiñón             223    1297,27      24707       1006000101   

            CAPITAL  POBLACION_CAPITAL HOJA_MTN25_ETRS89 LONGITUD_ETRS89  \
0  Alegría-Dulantzi               2823            0113-3     -2,51243731   
1           Amurrio               9188            0086-4     -3,00007326   
2            Ibarra                787            0087-4     -2,56540037   
3        Artziniega               1675            0086-1     -3,12791718   
4           Armiñón                100            0137-4     -2,87183475   

  LATITUD_ETRS89 ORIGENCOOR ALTITUD ORIGENALTITUD  
0    42,83981158       Mapa     568           MDT  
1    43,05427776       Mapa     219           MDT  
2    43,05119653       Mapa     333           MDT  
3    43,12084358       Mapa     210           MDT  
4    42,72326199       Mapa     467           MDT

In [52]:
# Obtengo el código del INE en esta tabla que es la que utiliza AEMET. Los nombres no son iguales lo que podría dar problemas
df_munic_ine.head()

CODAUTO CPRO CMUN DC               NOMBRE
0      16   01  051  3  Agurain/Salvatierra
1      16   01  001  4     Alegría-Dulantzi
2      16   01  002  9              Amurrio
3      16   01  049  3                Añana
4      16   01  003  5              Aramaio

In [53]:
df_munic_ine.shape

(8131, 5)

In [54]:
df_munic_ine.columns.values

array(['CODAUTO', 'CPRO', 'CMUN', 'DC', 'NOMBRE'], dtype=object)

In [55]:
# Concateno y creo COD_AEMET a partir de CPRO y CMUN
df_munic_ine["COD_AEMET"] = df_munic_ine["CPRO"].astype(str) + df_munic_ine["CMUN"].astype(str)

In [56]:
# Hago un join entre las dos tablas anteriores por COD_AEMET
df_munic_merge = df_munic_ine.merge(df_munic_tot, on="COD_AEMET", how='outer')

In [57]:
df_munic_merge.shape

(8131, 24)

In [58]:
df_munic_merge = df_munic_merge.sort_values(by = "COD_AEMET").reset_index(drop=True, inplace=False)

In [59]:
df_munic_merge.head()

CODAUTO CPRO CMUN DC            NOMBRE COD_AEMET     COD_INE   ID_REL  \
0      16   01  001  4  Alegría-Dulantzi     01001  1001000000  1010014   
1      16   01  002  9           Amurrio     01002  1002000000  1010029   
2      16   01  003  5           Aramaio     01003  1003000000  1010035   
3      16   01  004  0        Artziniega     01004  1004000000  1010040   
4      16   01  006  6           Armiñón     01006  1006000000  1010066   

   COD_GEO  COD_PROV  ... PERIMETRO COD_INE_CAPITAL           CAPITAL  \
0     1010         1  ...     35069      1001000101  Alegría-Dulantzi   
1     1020         1  ...     65652      1002000201           Amurrio   
2     1030         1  ...     41775      1003000601            Ibarra   
3     1040         1  ...     22886      1004000101        Artziniega   
4     1060         1  ...     24707      1006000101           Armiñón   

  POBLACION_CAPITAL  HOJA_MTN25_ETRS89  LONGITUD_ETRS89 LATITUD_ETRS89  \
0              2823             0113-3      -2,51243731    42,83981158   
1              9188             0086-4      -3,00007326    43,05427776   
2               787             0087-4      -2,56540037    43,05119653   
3              1675             0086-1      -3,12791718    43,12084358   
4               100             0137-4      -2,87183475    42,72326199   

   ORIGENCOOR ALTITUD ORIGENALTITUD  
0        Mapa     568           MDT  
1        Mapa     219           MDT  
2        Mapa     333           MDT  
3        Mapa     210           MDT  
4        Mapa     467           MDT  

[5 rows x 24 columns]

In [60]:
df_munic_merge.columns.values

array(['CODAUTO', 'CPRO', 'CMUN', 'DC', 'NOMBRE', 'COD_AEMET', 'COD_INE',
       'ID_REL', 'COD_GEO', 'COD_PROV', 'PROVINCIA', 'NOMBRE_ACTUAL',
       'POBLACION_MUNI', 'SUPERFICIE', 'PERIMETRO', 'COD_INE_CAPITAL',
       'CAPITAL', 'POBLACION_CAPITAL', 'HOJA_MTN25_ETRS89',
       'LONGITUD_ETRS89', 'LATITUD_ETRS89', 'ORIGENCOOR', 'ALTITUD',
       'ORIGENALTITUD'], dtype=object)

In [61]:
sum(df_munic_merge['NOMBRE'] != df_munic_merge['NOMBRE_ACTUAL'])

623

Al haber 623 municipios con nombres distintos, decidimos utilizar los más actualizados y sustituirlo en incendios_fin

In [62]:
# Quito variables que no aportan nada
df_munic_merge = df_munic_merge[['COD_AEMET', 'COD_PROV', 'PROVINCIA', 'NOMBRE', 'POBLACION_MUNI', 'SUPERFICIE',
                                 'LONGITUD_ETRS89', 'LATITUD_ETRS89', 'ALTITUD']]

In [63]:
# Renombramos las variables para que coincidan con las variables del dataset incendios_fin
df_munic_merge = df_munic_merge.rename(columns={'COD_PROV':'idprovincia',
                                                'PROVINCIA':'provincia',
                                                'NOMBRE':'municipio',
                                                'POBLACION_MUNI':'poblacion_muni',
                                                'SUPERFICIE':'superficie_muni',
                                                'LONGITUD_ETRS89':'longitude',
                                                'LATITUD_ETRS89':'latitude',
                                                'ALTITUD':'altitud'})

In [64]:
df_munic_merge.head()

COD_AEMET  idprovincia    provincia         municipio  poblacion_muni  \
0     01001            1  Araba/Álava  Alegría-Dulantzi            2935   
1     01002            1  Araba/Álava           Amurrio           10264   
2     01003            1  Araba/Álava           Aramaio            1442   
3     01004            1  Araba/Álava        Artziniega            1800   
4     01006            1  Araba/Álava           Armiñón             223   

  superficie_muni    longitude     latitude altitud  
0       1994,5872  -2,51243731  42,83981158     568  
1         9649,53  -3,00007326  43,05427776     219  
2         7325,84  -2,56540037  43,05119653     333  
3         2728,73  -3,12791718  43,12084358     210  
4         1297,27  -2,87183475  42,72326199     467

Agregamos otras columnas como comunidad, zona, zona metereológica...

In [67]:
com_prov = pd.read_csv("C:/Users/saulo/Documents/R/data/com_prov.csv",
                       sep=",",
                       engine='python',
                       dtype = {'idcomunidad': str})

In [68]:
com_prov.head()

comunidad provincia_x          zona idcomunidad  idprovincia
0  Andalucía     Almería  Mediterráneo           1            4
1  Andalucía       Cádiz  Mediterráneo           1           11
2  Andalucía     Córdoba  Mediterráneo           1           14
3  Andalucía     Granada  Mediterráneo           1           18
4  Andalucía      Huelva  Mediterráneo           1           21

In [69]:
com_prov = com_prov[['comunidad', 'zona', 'idcomunidad', 'idprovincia']]

In [70]:
df_munic_merge_com = df_munic_merge.merge(com_prov, on = "idprovincia", how = 'left')

In [71]:
df_munic_merge_com.head()

COD_AEMET  idprovincia    provincia         municipio  poblacion_muni  \
0     01001            1  Araba/Álava  Alegría-Dulantzi            2935   
1     01002            1  Araba/Álava           Amurrio           10264   
2     01003            1  Araba/Álava           Aramaio            1442   
3     01004            1  Araba/Álava        Artziniega            1800   
4     01006            1  Araba/Álava           Armiñón             223   

  superficie_muni    longitude     latitude altitud   comunidad      zona  \
0       1994,5872  -2,51243731  42,83981158     568  País Vasco  Noroeste   
1         9649,53  -3,00007326  43,05427776     219  País Vasco  Noroeste   
2         7325,84  -2,56540037  43,05119653     333  País Vasco  Noroeste   
3         2728,73  -3,12791718  43,12084358     210  País Vasco  Noroeste   
4         1297,27  -2,87183475  42,72326199     467  País Vasco  Noroeste   

  idcomunidad  
0          16  
1          16  
2          16  
3          16  
4          16

In [72]:
df_zonas_met = pd.read_csv("C:/Users/saulo/Documents/R/data/zona_meteo.csv",
                       sep=";",
                       engine='python',
                       dtype = {'COD_AEMET': object})

In [73]:
df_zonas_met.shape

(8131, 6)

In [74]:
df_zonas_met.head()

idprovincia    provincia  id_zona_meteo          zona_meteo COD_AEMET  \
0            1  Araba/Álava         750102     Llanada alavesa     01001   
1            1  Araba/Álava         750101  Cuenca del Nervión     01002   
2            1  Araba/Álava         750101  Cuenca del Nervión     01003   
3            1  Araba/Álava         750101  Cuenca del Nervión     01004   
4            1  Araba/Álava         750102     Llanada alavesa     01006   

          municipio  
0  Alegría-Dulantzi  
1           Amurrio  
2           Aramaio  
3        Artziniega  
4           Armiñón

In [75]:
df_zonas_met = df_zonas_met[['COD_AEMET', 'id_zona_meteo', 'zona_meteo']]

In [76]:
df_munic_merge_fin = df_munic_merge_com.merge(df_zonas_met, on = "COD_AEMET", how = 'inner')

In [77]:
df_munic_merge_fin

COD_AEMET  idprovincia    provincia              municipio  \
0        01001            1  Araba/Álava       Alegría-Dulantzi   
1        01002            1  Araba/Álava                Amurrio   
2        01003            1  Araba/Álava                Aramaio   
3        01004            1  Araba/Álava             Artziniega   
4        01006            1  Araba/Álava                Armiñón   
...        ...          ...          ...                    ...   
8126     50901           50     Zaragoza                   Biel   
8127     50902           50     Zaragoza               Marracos   
8128     50903           50     Zaragoza  Villamayor de Gállego   
8129     51001           51        Ceuta                  Ceuta   
8130     52001           52      Melilla                Melilla   

      poblacion_muni superficie_muni     longitude     latitude altitud  \
0               2935       1994,5872   -2,51243731  42,83981158     568   
1              10264         9649,53   -3,00007326  43,05427776     219   
2               1442         7325,84   -2,56540037  43,05119653     333   
3               1800         2728,73   -3,12791718  43,12084358     210   
4                223         1297,27   -2,87183475  42,72326199     467   
...              ...             ...           ...          ...     ...   
8126             155      13072,6145   -0,93657238  42,38756817     755   
8127              84       1692,4833   -0,77641575  42,09091716     413   
8128            2745       8936,4544    -0,7705331  41,68419824     232   
8129           84202            1987   -5,30675127  35,88810209      27   
8130           87076         1424,28  -2,937088836  35,29220967       2   

       comunidad          zona idcomunidad  id_zona_meteo  \
0     País Vasco      Noroeste          16         750102   
1     País Vasco      Noroeste          16         750101   
2     País Vasco      Noroeste          16         750101   
3     País Vasco      Noroeste          16         750101   
4     País Vasco      Noroeste          16         750102   
...          ...           ...         ...            ...   
8126      Aragón      Interior           2         625001   
8127      Aragón      Interior           2         625001   
8128      Aragón      Interior           2         625003   
8129       Ceuta  Mediterráneo          18         785101   
8130     Melilla  Mediterráneo          19         795201   

                       zona_meteo  
0                 Llanada alavesa  
1              Cuenca del Nervión  
2              Cuenca del Nervión  
3              Cuenca del Nervión  
4                 Llanada alavesa  
...                           ...  
8126     Cinco Villas de Zaragoza  
8127     Cinco Villas de Zaragoza  
8128  Ribera del Ebro de Zaragoza  
8129                        Ceuta  
8130                      Melilla  

[8131 rows x 14 columns]

In [78]:
df_munic_merge_fin.shape

(8131, 14)

In [79]:
df_munic_merge_fin.isnull().sum()

COD_AEMET          0
idprovincia        0
provincia          0
municipio          0
poblacion_muni     0
superficie_muni    0
longitude          0
latitude           0
altitud            0
comunidad          0
zona               0
idcomunidad        0
id_zona_meteo      0
zona_meteo         0
dtype: int64

In [80]:
df_munic_merge_fin.columns.values

array(['COD_AEMET', 'idprovincia', 'provincia', 'municipio',
       'poblacion_muni', 'superficie_muni', 'longitude', 'latitude',
       'altitud', 'comunidad', 'zona', 'idcomunidad', 'id_zona_meteo',
       'zona_meteo'], dtype=object)

In [81]:
# Quito variables que no aportan nada y las ordeno
df_munic_merge_fin = df_munic_merge_fin[['latitude', 'longitude', 'idcomunidad', 'comunidad', 'idprovincia', 'provincia',
                                         'COD_AEMET', 'municipio', 'zona', 'altitud', 'poblacion_muni', 'superficie_muni', 
                                         'id_zona_meteo', 'zona_meteo']]

In [82]:
df_munic_merge_fin.head()

latitude    longitude idcomunidad   comunidad  idprovincia    provincia  \
0  42,83981158  -2,51243731          16  País Vasco            1  Araba/Álava   
1  43,05427776  -3,00007326          16  País Vasco            1  Araba/Álava   
2  43,05119653  -2,56540037          16  País Vasco            1  Araba/Álava   
3  43,12084358  -3,12791718          16  País Vasco            1  Araba/Álava   
4  42,72326199  -2,87183475          16  País Vasco            1  Araba/Álava   

  COD_AEMET         municipio      zona altitud  poblacion_muni  \
0     01001  Alegría-Dulantzi  Noroeste     568            2935   
1     01002           Amurrio  Noroeste     219           10264   
2     01003           Aramaio  Noroeste     333            1442   
3     01004        Artziniega  Noroeste     210            1800   
4     01006           Armiñón  Noroeste     467             223   

  superficie_muni  id_zona_meteo          zona_meteo  
0       1994,5872         750102     Llanada alavesa  
1         9649,53         750101  Cuenca del Nervión  
2         7325,84         750101  Cuenca del Nervión  
3         2728,73         750101  Cuenca del Nervión  
4         1297,27         750102     Llanada alavesa

In [83]:
df_munic_merge_fin.to_csv('municipiosINE.csv', index = False)

### Descomponer dataset de predicciones 

In [ ]:
# obtenemos la lista con todos los códigos del INE de los municipios de España para poder enviarlos a la API de AEMET
municipios = df_munic_merge["COD_AEMET"]

In [ ]:
print(type(municipios))

In [ ]:
municipios = municipios.tolist()

In [ ]:
print(type(municipios))

In [ ]:
columns_name = ['temperatura', 'estadoCielo', 'humedadRelativa', 'uvMax', 'sensTermica', 'viento',
                 'cotaNieveProv', 'fecha', 'rachaMax', 'probPrecipitacion','municipio', 'provincia']

In [ ]:
df_total = pd.DataFrame(columns = columns_name)

Con el siguiente código llamamos a la API de AEMET y obtenemos un JSON con la predicción para los próximos 7 días para un pueblo o todos los que le pasemos en la lista municipios (en este caso sólo le paso uno).  

Se obtiene un fichero JSON con diccionarios y listas que tenemos que desanidar para obtener el valor máximo de cada variable para cada uno de los próximos 7 días.  

Las listas las desanidamos con explode y obtenemos el valor máximo. Para los diccionarios utilizamos json_normalize.  

Finalmente unimos los dataframes resultantes y le agregamos la información del municipio proveniente de la tabla df_munic_merge.

In [ ]:
columns_name = ['temperatura', 'estadoCielo', 'humedadRelativa', 'uvMax', 'sensTermica', 'viento',
                 'cotaNieveProv', 'fecha', 'rachaMax', 'probPrecipitacion','municipio', 'provincia']
df_total = pd.DataFrame(columns = columns_name)
for j in municipios[8126:8127]: 
    b = obtenerUrl(j)
    a = obtenerDatos(b)
    prediccion = a['prediccion'][0].get('dia')
    municipio = a['nombre']
    provincia = a['provincia']
    for i in range(len(prediccion)):
        fila_dia = pd.DataFrame(prediccion[i].items()).T
        fila_dia.columns = fila_dia.iloc[0]
        fila_dia.drop([0],inplace=True)
        fila_dia_exp = fila_dia.explode("probPrecipitacion") # probabilidad de precipitación
        lista_prob = []
        for k in range(len(fila_dia_exp["probPrecipitacion"])):
            lista_prob.append(fila_dia_exp.iloc[k,0]['value'])
        fila_dia["probPrecipitacion"] = max(lista_prob)
        fila_dia_exp = fila_dia.explode("cotaNieveProv") # cota de nieve en la provincia
        lista_cota = []
        for k in range(len(fila_dia_exp["cotaNieveProv"])):
            if fila_dia_exp.iloc[k,1]['value'] == "":
                fila_dia_exp.iloc[k,1]['value'] = 0
            lista_cota.append(fila_dia_exp.iloc[k,1]['value'])
        fila_dia["cotaNieveProv"] = max(lista_cota)
        fila_dia_exp = fila_dia.explode("viento") # velocidad y dirección del viento
        lista_dir = []
        lista_vel = []
        for k in range(len(fila_dia_exp["viento"])):
            if fila_dia_exp.iloc[k,3]['velocidad'] == 0:
                fila_dia_exp.iloc[k,3]['direccion'] = 'C'
            lista_dir.append(fila_dia_exp.iloc[k,3]['direccion'])
            lista_vel.append(fila_dia_exp.iloc[k,3]['velocidad'])
        fila_dia["velmedia"] = max(lista_vel)
        fila_dia["dir"] = lista_dir[lista_vel.index(max(lista_vel))]
        fila_dia_exp = fila_dia.explode("rachaMax") # racha máxima viento
        lista_vie = []
        for k in range(len(fila_dia_exp["rachaMax"])):
            if fila_dia_exp.iloc[k,4]['value'] == "":
                fila_dia_exp.iloc[k,4]['value'] = 0
            lista_vie.append(fila_dia_exp.iloc[k,4]['value'])        
        lista_vie = [int(i) for i in lista_vie]        
        fila_dia["rachaMax"] = max(lista_vie)        
        fila_dia.insert(9,'municipio',municipio[0])
        fila_dia.insert(10,'provincia',provincia[0])
        df_total = pd.concat([df_total,fila_dia])

In [ ]:
df_total

In [ ]:
df_normal = json_normalize(a['prediccion'][0].get('dia'))

In [ ]:
df_normal = df_normal[["fecha", "temperatura.maxima", "temperatura.minima", "sensTermica.maxima",
                       "sensTermica.minima", "humedadRelativa.maxima", "humedadRelativa.minima"]]

In [ ]:
df_normal

In [ ]:
df_final = df_normal.merge(df_total, on="fecha", how='outer')

In [ ]:
df_final = df_final.drop(["temperatura", "estadoCielo", "humedadRelativa", "sensTermica", "viento"], axis=1)

In [ ]:
df_final.columns

In [ ]:
df_final.columns = ["fecha", "tmax", "tmin", "senTmax", "senTmin", "hRmax", "hRmin", "uvMax", "cotaNieve", "racha",
                   "probPrecipitacion", "municipio", "provincia", "velmedia", "dir"]

In [ ]:
df_final

In [ ]:
df_completo = df_munic_merge.merge(df_final, left_on="NOMBRE ", right_on='municipio', how='right')

In [ ]:
df_completo.columns

In [ ]:
df_completo = df_completo[['COD_PROV', 'PROVINCIA', 'NOMBRE_ACTUAL', 'NOMBRE ', 'LONGITUD_ETRS89', 'LATITUD_ETRS89', 'ALTITUD',
       'fecha', 'tmax', 'tmin', 'senTmax', 'senTmin', 'hRmax', 'hRmin', 'uvMax', 'cotaNieve', 'racha', 'probPrecipitacion',
       'velmedia', 'dir']]

In [ ]:
df_completo